signer/xpi/cose.go

```diff
@@ -0,0 +1,259 @@
+func isSupportedCOSEAlgValue(algValue interface{}) bool {
+	return algValue == cose.PS256.Value || algValue == cose.ES256.Value || algValue == cose.ES384.Value || algValue == cose.ES512.Value
+}
+
+// isValidCOSESignature checks whether a COSE signature is a valid for XPIs
+func isValidCOSESignature(sig cose.Signature) (eeCert *x509.Certificate, resultErr error) {
+	if len(sig.Headers.Unprotected) != 0 {
+		resultErr = fmt.Errorf("XPI COSE Signature must have an empty Unprotected Header")
+		return
+	}
+
+	if len(sig.Headers.Protected) != 2 {
+		resultErr = fmt.Errorf("XPI COSE Signature must have exactly two Protected Headers")
+		return
+	}
+	algValue, ok := sig.Headers.Protected[1] // 1 is the compressed key for "alg"
+	if !ok {
+		resultErr = fmt.Errorf("XPI COSE Signature must have alg in Protected Headers")
+		return
+	}
+	if !isSupportedCOSEAlgValue(algValue) {
+		resultErr = fmt.Errorf("XPI COSE Signature must have alg %+v is not supported", algValue)
+		return
+	}
+
+	kidValue, ok := sig.Headers.Protected[4] // 4 is the compressed key for "kid"
+	if !ok {
+		resultErr = fmt.Errorf("XPI COSE Signature must have kid in Protected Headers")
+		return
+	}
+	kidBytes, ok := kidValue.([]byte)
+	if !ok {
+		resultErr = fmt.Errorf("XPI COSE Signature kid value is not bytes")
+		return
+	}
+
+	eeCert, err := x509.ParseCertificate(kidBytes) // eeCert
+	if err != nil {
+		resultErr = errors.Wrapf(err, "XPI COSE Signature kid must decode to a parseable X509 cert")
+		return
+	}
+	return
+}
+
+// isValidCOSEMessage checks whether a COSE SignMessage is a valid for
+// XPIs and returns parsed intermediate and end entity certs
+func isValidCOSEMessage(msg cose.SignMessage) (intermediateCerts, eeCerts []*x509.Certificate, resultErr error) {
+	if msg.Payload != nil {
+		resultErr = fmt.Errorf("Expected SignMessage payload to be nil, but got %+v", msg.Payload)
+		return
+	}
+	if len(msg.Headers.Unprotected) != 0 {
+		resultErr = fmt.Errorf("Expected SignMessage Unprotected headers to be empty, but got %+v", msg.Headers.Unprotected)
+		return
+	}
+
+	if len(msg.Headers.Protected) != 1 {
+		resultErr = fmt.Errorf("Expected SignMessage Protected headers must contain one value, but got %d", len(msg.Headers.Protected))
+		return
+	}
+	kidValue, ok := msg.Headers.Protected[4] // 4 is the compressed key for "kid"
+	if !ok {
+		resultErr = fmt.Errorf("Expected SignMessage must have kid in Protected Headers")
+		return
+	}
+	// check that all kid values are bytes and decode into certs
+	kidArray, ok := kidValue.([]interface{})
+	if !ok {
+		resultErr = fmt.Errorf("Expected SignMessage Protected Headers kid value to be an array got %+v with type %T", kidValue, kidValue)
+		return
+	}
+	for i, cert := range kidArray {
+		certBytes, ok := cert.([]byte)
+		if !ok {
+			resultErr = fmt.Errorf("Expected SignMessage Protected Headers kid value %d to be a byte slice got %+v with type %T", i, cert, cert)
```

Comment:
do we want to return `%+v` in standard errors? it feel more like debugging details to me, and would probably make logs massive and hard to parse in production.